backend/api/deck.go

```diff
@@ -41,38 +44,206 @@ func (a *API) getDeckHandler(c *gin.Context) {
+		// TODO: some validation here
+
+		cards = append(cards, val)
+		if len(val) > longestCard {
+			longestCard = len(val)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	cardIdxCountMap := make(map[string]int, len(cards)) // map[card]count
+	uniqueCardList := make([]string, 0, len(cards))     // just keep copies of our strs
+
+	// now that we know which cards exist, we can read the indexes
+	err = readDelimitedBytes(parts[0], []byte(","), func(val []byte) error {
+		// compiler will optimize this string conversion out, see `cloneString` in atoi.go
+		cardIdxVal, err := strconv.ParseInt(string(val), 10, 64)
+		if err != nil {
+			return err
+		}
+
+		if cardIdxVal < 0 || cardIdxVal >= int64(len(cards)) {
+			return errors.New("card index out of bounds")
+		}
+
+		card := parseCardBytes(cards[cardIdxVal])
+		// fmt.Printf("%d - %s - %s\n", cardIdxVal, string(card), string(cards[cardIdxVal]))
+
+		if len(card) == 0 {
+			return nil
+		}
+
+		// no guarentee of not copying here in map lookup, so unsafe it out. Is valid as long as `decompressed` is still valid and static
+		cardStr := unsafe.String(&card[0], len(card))
+
+		ct, ok := cardIdxCountMap[cardStr]
+		if !ok {
+			uniqueCardList = append(uniqueCardList, cardStr)
+		}
+		cardIdxCountMap[cardStr] = ct + 1
+
+		return nil
+	})
+	if err != nil {
+		return err
 	}
 
-	slices.SortFunc(keys, func(i, j string) bool {
+	// may allocate some extra space in some cases, but we will shrink after we are done formatting
+	d.buf = make([]byte, 0, len(cardIdxCountMap)*longestCard)
+
+	// put ascender's bane first
+	slices.SortFunc(uniqueCardList, func(i, j string) bool {
 		if i == "Ascender's Bane" {
 			return false
 		}
 		if j == "Ascender's Bane" {
 			return true
 		}
+
 		return i < j
 	})
 
-	result := strings.Builder{}
-	for _, k := range keys {
-		result.WriteString(k)
-		result.WriteString(" x")
-		result.WriteString(fmt.Sprint(d[k]))
-		result.WriteString("\n")
+	for _, card := range uniqueCardList {
+		// will just use underlying bytes and not do cast
+		d.buf = append(d.buf, []byte(card)...)
+
+		cardCount, ok := cardIdxCountMap[card]
+		if !ok {
+			return errors.New("card not found")
+		}
+
+		// fmt - "$card x$count\n"
+		if cardCount > 0 {
+			d.buf = append(d.buf, ' ', 'x')
+			d.buf = strconv.AppendInt(d.buf, int64(cardCount), 10)
+			d.buf = append(d.buf, '\n')
+		}
+	}
+
+	return nil
+}
+
+func decompressBytes(s []byte) ([]byte, error) {
+	parts := bytes.Split(s, []byte("||"))
+	if len(parts) < 2 {
+		return nil, errors.New("invalid deck")
+	}
+
+	compressionDict := bytes.Split(parts[0], []byte("|"))
+	text := parts[1]
+
+	for i := len(compressionDict) - 1; i >= 0; i-- {
+		word := compressionDict[i]
+		// TODO: this is the source of lots of allocs and CPU cycles, probably no need for regexp here
+		text = compressionWildcardRegex[i].ReplaceAll(text, word)
+	}
+
+	return text, nil
+}
+
+type delimCB func(val []byte) error
```

Comment:
Nit: for type names especially could we type out fuller names? I think this stands for callback, but it took me a bit to guess.